### DISCUSSION TITLE
Best way to exit from nested scripts

### DISCUSSION THREAD

**Discussion by srett** [ipxe/ipxe#341] [Q&A] [answered]
I'm generating ipxe menus via PHP server-side where every menu-entry just chains to another script snippet. This is pretty neat and modular, you can easily nest menus etc. However, in general since I don't know if the snippet I chain to could want to return to the previous menu, I'm not using `chain --replace`. Now, if I eventually arrive at a script snippet that wants to exit from iPXE (for local boot), I have a problem: doing `exit` simply returns to the previous ipxe script and continues execution there. Ideally, there would be something like `exit --recursive`, but what's the next best thing to do? I was thinking about settling on a special variable that a script can set, and the caller will just issue an `exit` if it's set after `chain` returns.
```
set exit_val ${}
chain -a <selected_entry_url> ||
iseq ${exit_val} ${} || exit ${exit_val}
```
This should work recursively, but everybody using this system has to play along so it feels rather clumsy, or is there a better way to achieve this?

**Comment by mcb30** [comment, marked as answer]
It's probably a good idea to avoid that kind of recursion.  Apart from any other considerations, you are likely to run out of stack space after a few levels.

I'd suggest using `--replace` and keeping track of the return point (rather than relying on `exit` to return to a previous script).  Alternatively, if you can refactor your menus into a single script and use `goto` for navigation, that would also avoid the problem (and would avoid the need to handle failures of the `chain` command).

**Reply by srett** [reply]
> It's probably a good idea to avoid that kind of recursion. Apart from any other considerations, you are likely to run out of stack space after a few levels.

What are the actual constraints here? I'd have assumed that at least with EFI, this shouldn't be a problem. Realistically it's 2-3 levels of menus. Some of the script snippets for example set variables that get eventually passed to the kernel command line by another menu entry. We use this setup for HW compatibility testing, to boot different kernels, have menu entries to enable/disable certain kernel command line options, boot into iSCSI Windows installs, and occasionally, local boot. For BIOS, `sanboot --no-describe` was the most reliable solution for local boot.

> I'd suggest using `--replace` and keeping track of the return point (rather than relying on `exit` to return to a previous script).

How would I do this over multiple levels? Setting a variable would lose the previous return point if I'm already in a sub-menu. I could pass it along as URL parameters in array form (multiple `&return[]=`) but I don't think it will become easier that way.

> Alternatively, if you can refactor your menus into a single script and use `goto` for navigation, that would also avoid the problem (and would avoid the need to handle failures of the `chain` command).

This is where we started out, it had two main problems: If you make changes to any script snippet, you have to remember to reload the menu if you're already in it. Not a deal breaker, but you forget it often enough for it to become annoying. The other problem was label collisions in script snippets, and error handling. As soon as there's an error in a snippet that's not handled, the whole thing exits. We're about a dozen people accessing this system, so it was hard to enforce proper error handling etc. `chain ... ||` was an easy way to get this under control. But now that I think about it some more, just blindly appending `|| goto error` to every single line of user supplied script that doesn't already end in it shouldn't break anything.... And maybe it's not too much effort to write a simple parser that finds all labels and gotos in those snippets and mangles them to be sure they are unique. Hmm...

I guess iPXE was never intended for huge menu trees, but it it has made testing in general much easier and faster. It's very flexible.

**Reply by mcb30** [reply]
> > It's probably a good idea to avoid that kind of recursion. Apart from any other considerations, you are likely to run out of stack space after a few levels.
> What are the actual constraints here? I'd have assumed that at least with EFI, this shouldn't be a problem. Realistically it's 2-3 levels of menus.

The constraints are not well-defined, but 2-3 levels should be fine.  On BIOS, the stack space is internal and fixed at build time.  On UEFI, the stack is externally provided and the UEFI specification provides no guarantees as to its size (or as to what happens when you underrun the stack).

> > I'd suggest using `--replace` and keeping track of the return point (rather than relying on `exit` to return to a previous script).
> 
> How would I do this over multiple levels? Setting a variable would lose the previous return point if I'm already in a sub-menu. I could pass it along as URL parameters in array form (multiple `&return[]=`) but I don't think it will become easier that way.

Can submenus be entered from more than one place?  If each submenu can be entered only from one source, then each submenu has a well-defined return point.  (Of course, this will depend entirely on how you have your menus structured.)

> This is where we started out, it had two main problems: If you make changes to any script snippet, you have to remember to reload the menu if you're already in it. Not a deal breaker, but you forget it often enough for it to become annoying. The other problem was label collisions in script snippets, and error handling. As soon as there's an error in a snippet that's not handled, the whole thing exits. We're about a dozen people accessing this system, so it was hard to enforce proper error handling etc. `chain ... ||` was an easy way to get this under control. But now that I think about it some more, just blindly appending `|| goto error` to every single line of user supplied script that doesn't already end in it shouldn't break anything.... And maybe it's not too much effort to write a simple parser that finds all labels and gotos in those snippets and mangles them to be sure they are unique. Hmm...
> 
> I guess iPXE was never intended for huge menu trees, but it it has made testing in general much easier and faster. It's very flexible.

Yeah, you're hitting the limits of what it was ever designed to do.  I'm glad it's proving useful.  :slightly_smiling_face: 

One possible compromise could be to have a single "outer" script that invokes the menu script and handles any unintended return from the menu script.  This would avoid the need for the menu script itself to have error handling on every single line.

**Reply by srett** [reply]
> > I guess iPXE was never intended for huge menu trees, but it it has made testing in general much easier and faster. It's very flexible.
> 
> Yeah, you're hitting the limits of what it was ever designed to do. I'm glad it's proving useful. slightly_smiling_face
> 
> One possible compromise could be to have a single "outer" script that invokes the menu script and handles any unintended return from the menu script. This would avoid the need for the menu script itself to have error handling on every single line.

That sounds like an easy pragmatic solution for now. Will think some more about the general approach. Thanks for all the input!
